### ISSUE TITLE
Proper handling of a corrupted cache

### ISSUE THREAD

**Issue by doraskayo** [flatpak/flatpak#554]
This morning my operating system crashed during an ordinary `flatpak update`, and Flatpak's cache apparently got corrupted. I know now, but it certainly wasn't easy to figure out. From that point on, any time I tried to install or upgrade either an application or a runtime, the CLI helper gave me the following vague error:
```
error: Invalid UTF-8
```
GNOME Software was more generous and gave me the following when trying to update, but doesn't mention Flatpak at all:
```
Failed to get AppStream metadata: Invalid UTF-8
```
And that's it. No other helpful information at all. It took me all morning to figure out what caused this, after trying to uninstall and reinstall everything related to Flatpak in my system, and even renaming `/var/lib/flatpak` to force Flatpak to regenerate its data did absolutely nothing. Only after using the [`strace`](https://linux.die.net/man/1/strace) to trace Flatpak's file access when attempting to install an app, I noticed that files in `/home/dor/.local/share/flatpak/system-cache` were accessed just before the error appears. Renaming that folder finally fixed my issue and allowed installing and updating again.

In that little journey I came to a few worrisome conclusions:
* Flatpak can't gracefully handle a corrupted cache, and requires manual intervention.
* Flatpak doesn't provide enough information in the error it shows when its cache is corrupted, or a clear way to try to get around the issue.
* Flatpak doesn't provide a command to manually clean its cache, or an option to skip the cache when executing commands.
* There is no documentation (that I could find) about the directories Flatpak uses to store its cache.
* There is no documentation (that I could find) about a proper method to completely remove Flatpak from the system, including the various folders and files it creates during its operation.
* Flatpak can potentially break system updates, since GNOME Software will not allow them to continue when Flatpak errors out.

At least most of these issues should be handled in a better way, so that less people find themselves scratching their heads wondering why their system refuses to update.

**Comment by alexlarsson** [comment]
Yeah. Its a bit complicated because we *do* want to reuse old cache directories, otherwise we'll have to start downloading updates from scratch on intermittent failures. But yeah, clearly we should report errors better and recover better from such failures.

**Comment by johan-bjareholt** [comment]
I had the same issue, but I did not have a `~/.local/share/flatpak/system-cache` folder.

The folder I needed to (re)move was the `.flatpak-builder/cache` folder.
